### PR TITLE
dialects: Add mpi.gather operation with lowering

### PR DESCRIPTION
**Pull request by AntonLydike**

### CODE REVIEW

**Review by georgebisbas — APPROVED** [review]
Cool

**Review by PapyChacal — APPROVED** [review]
:tada:

File: xdsl/dialects/mpi.py [hunk 1]
```diff
@@ -775,6 +775,63 @@ def get(req: SSAValue | Operation):
...
+                    int root,
+                    MPI_Comm comm)
+
+     - sendbuf, sendcount, sendtype: info on the buffer to be sent
+     - recvbuf, recvcount, recvtype: info on the gather buffer
+     - root: the rank that receives all the data
+     - comm: the communicator to use
+
+    Note: recvcount * sizeof(recvtype) == sendcount * sizeof(sendtype) * N
+    where N is the number of nodes participating in the gather.
```
Comment (webmiche) [inline xdsl/dialects/mpi.py:794]: Can you somehow enforce this in a verifier? I am guessing that you are missing N, right?

Comment (AntonLydike) [reply xdsl/dialects/mpi.py:794]: We are missing N, and it's all SSA values right?

Comment (webmiche) [reply xdsl/dialects/mpi.py:794]: Ah, true. Why is there no facepalm on git? 🤔

**Pull request merged**